[RFC] Use config.injections.aliases to inject a language

stsewd opened the pull request:
Closes https://github.com/nvim-treesitter/nvim-treesitter/issues/812
Supersedes https://github.com/nvim-treesitter/nvim-treesitter/pull/1190

stsewd commented on lua/nvim-treesitter/parsers.lua:
```diff
@@ -34,6 +34,9 @@ list.javascript = {
     files = { "src/parser.c", "src/scanner.c" },
   },
+  injections = {
+    pattern = '\\c^js$'
```
Should we have an option instead of manually setting this? Like `injections.ignore_case`

stsewd commented:
There are more languages that can be updated, but first wanted to have your feedback in the final implementation

theHamsta commented on lua/nvim-treesitter/query_predicates.lua:
```diff
@@ -85,5 +86,34 @@ query.add_predicate('has-type?', function(match, pattern, bufnr, pred)
+  local configs = parsers.get_parser_configs()
+  local language = default
+  for lang in pairs(configs) do
+    if text == lang then
```
why do a reverse look up? Going through all languages and all available regexes? Wouldn't one table with ["py"] = "python" faster and less distributed to configure? Normalization like lowercase could be done before.

theHamsta replied:
Maybe take an already existing table: https://github.com/tjdevries/plenary.nvim/blob/9bd408ba679d1511e269613af840c5019de59024/data/plenary/filetypes/base.lua#L1

kyazdani42 replied:
i agree, a table will speed up the directive impl

vigoux replied:
Agreed on the table for lookup here too

steelsojka replied:
👍

theHamsta replied:
A CI jobs could also help generate parts of the table via their ftdetect table. I imagine the mapping extension to language could be handy for example in Markdown.

stsewd commented:
So, changed the implementation to use just a table, I started this as a regex to make it more general, but didn't find any language that could make use of such pattern. I added some aliases of languages I know. Let me know what you think of this implementation.

3nuc commented on queries/vue/injections.scm:
```diff
@@ -1,26 +1,12 @@
-)
+  (start_tag (attribute (quoted_attribute_value (attribute_value) @_lang))?)
+  (raw_text) @content)
+ (#inject! @_lang "css"))
```
Do the changes in this file have the below logic for Vue?
1. > When you encounter a `<style lang="scss">` tag, inject scss
2. > When you encounter a `<style>` (no lang attribute), inject css

I see this PR removes

```
 (raw_text) @scss)
  (#match? @_lang "(scss|postcss|less)")
```
Doesn't it mean scss is never injected?

When I open your branch on a clean config (just nvim-treesitter) I get:
![image](https://user-images.githubusercontent.com/5359825/115974002-c0c8f580-a559-11eb-9d29-37d11e73394d.png)  
(the `$variableandor:` is scss specific syntax) 

The fix here would be to implement what is mentioned here: https://github.com/nvim-treesitter/nvim-treesitter/issues/1209#issuecomment-823878890

So if lang="scss", inject only scss, but if no lang attribute, inject css

I'd also keep the "inject scss for scss,postcss,less" (  (#match? @_lang "(scss|postcss|less)")) line for now

3nuc replied:
Is the situation with `<script lang="ts">` not similar? ie. typescript not injected

stsewd replied:
This change injects css if there isn't a lang tag, otherwise injects the language from the lang tag, if postcss or less is given, it will inject scss (they are defined as aliases).

In the adobe screenshot scss is being injected. The variables are white bc they are capture like so in https://github.com/stsewd/nvim-treesitter/blob/injection-regex/queries/scss/highlights.scm (and looks like there are some missing captures for `$` and `:`, you can check this by opening a `test.scss` file with that content.

stsewd replied:
looks like `:` is already captured, and the parser doesn't distinguish `$` from the var name.

3nuc replied:
:man_facepalming: it's all in the PR code, the aliases make sense. Sorry

tbh my knowledge about the project is rather limited and I don't want to keep bugging you with my potentially missed shots, so I'll just dump what I came across so you or the maintainers could also confirm it if it appears on their radar:

* if I check out to your branch and do `nvim index.vue`, I get
    `Error detected while processing FileType Autocommands for "*":
E5108: Error executing lua /usr/share/nvim/runtime/lua/vim/treesitter/lan
guagetree.lua:337: table index is nil`  
This doesn't happen if I just open `nvim` and then navigate to the file using telescope etc. If I revert back to master this issue disappears (this is reproduceable with a clean config like https://github.com/3nuc/tree-sitter-vue-issue-10/tree/stsewd, repro instructions are in issue 1209). This may be just some weird cached stuff on my side so do let me know if it works fine on your side. So perhaps this is why I'm running into the injection problems.

stsewd replied:
Nah, feel free to keep posting all errors you find. I wasn't able to replicate this with  https://github.com/3nuc/tree-sitter-vue-issue-10/blob/stsewd/index.vue, can you try modifying the file like adding a new line and saving it and opening it again and see if stills happens?

3nuc replied:
Yeah I still get this even if I change the file, save and repoen. Do you mean you couldn't replicate it by running index.vue on your config or that you cloned that repo with --recurse-submodules and ran ./run.sh from there?

stsewd replied:
I wasn't able to replicate with ./run.sh in that branch, yeah.

3nuc replied:
Let's just write it off as something wrong with my setup for now. If I manage to reproducce it on something squeaky clean like an empty VM with just neovim I'll let you know here.

elianiva replied:
FWIW you could use Nix and try to run Neovim inside a pure shell, it's easier than creating a VM in my opinion

DerekStride commented on lua/nvim-treesitter/parsers.lua:
```diff
@@ -557,6 +612,10 @@ function M.ft_to_lang(ft)
+function M.alias_to_lang(alias)
+  return injections_aliases[alias]
+end
+
```
Might be a good idea to add a function to allow user's to add custom alias' too.

```suggestion

function M.add_injection_alias(alias, lang)
  injection_aliases[alias] = lang
end
```

DerekStride commented on queries/ruby/injections.scm:
```diff
@@ -1 +1,6 @@
+
+(heredoc_body
+ (heredoc_content) @content
+ (heredoc_end) @_lang
+ (#inject! @_lang))
```
I'm not a huge fan of having to add this directive in order to trigger a hash lookup for the parser name. Is there a way to make this default behaviour when specifying an `@language` capture?

stsewd replied:
For that this would need to be ported upstream, since neovim does the injection. I'll try to look into that, I don't remember if neovim had access to the parser's configs

theHamsta commented on lua/nvim-treesitter/parsers.lua:
```diff
@@ -2,14 +2,24 @@ local api = vim.api
+      injections_aliases[alias] = name
+    end
+  end
+  injections_aliases[name] = name
```
do we really want the same problems as with `used_by` again. We could refactor to make the mapping `ft -> lang` and `alias -> lang` directly accessible so that the are directly configurable without any metatables and additional code magic.

steelsojka commented on lua/nvim-treesitter/query_predicates.lua:
```diff
@@ -85,5 +86,22 @@ query.add_predicate('has-type?', function(match, pattern, bufnr, pred)
+-- Inject a language based on the aliases from `config.injections.aliases`.
+-- Usage: (#inject! @_language [default])
+query.add_directive("inject!", function(match, pattern, bufnr, pred, metadata)
+  local match_id = pred[2]
```
Can we extract most of this logic to a function that can be used by other directives? We may have other directives that do more complex injection resolution, but still will need to do alias look ups.

Ontonator commented:
What is the status of this? I'm still having the issue mentioned in #1209 (although with Javascript/Typescript rather than CSS/SCSS), which was closed citing this as the future solution.

theHamsta commented:
> What is the status of this? I'm still having the issue mentioned in #1209 (although with Javascript/Typescript rather than CSS/SCSS), which was closed citing this as the future solution.

Preferably, this is addressed upstream in Neovim as there most logic for language injection is located there. Closing as this diverged a lot from our code base and that now that `:h vim.treesitter.add()` it would be a perfect fit to be able to set aliases for a language or (ab)use filetype.lua to detect a filetype by extension.
